client/auth/controller.js

```diff
@@ -25,14 +35,84 @@ module.exports = {
 	},
 
 	checkToken: function( context, next ) {
-		const loggedOutRoutes = [ '/login', '/oauth', '/start' ],
+		const loggedOutRoutes = [ '/login', '/oauth', '/start', '/authorize', '/api/oauth/token' ],
 			isValidSection = loggedOutRoutes.some( route => startsWith( context.path, route ) );
 
-		// Check we have an OAuth token, otherwise redirect to login page
+		if ( config( 'env' ) === 'desktop' ) {
+			// Check we have an OAuth token, otherwise redirect to login page
+			if ( OAuthToken.getToken() === false && ! isValidSection ) {
+				page( '/login' );
+			} else {
+				next();
+			}
+		}
+
+		// Check we have an OAuth token, otherwise redirect to auth page
 		if ( OAuthToken.getToken() === false && ! isValidSection ) {
-			page( '/login' );
+			page( '/authorize' );
 		} else {
 			next();
 		}
+	},
+
+	// This controller renders the API authentication screen
+	// for granting the app access to the user data using oauth
+	authorize: function() {
+		let authUrl;
+
+		if ( config( 'oauth_client_id' ) ) {
+			const oauthSettings = {
+				response_type: 'token',
+				client_id: config( 'oauth_client_id' ),
+				client_secret: 'n/a',
+				url: {
+					redirect: 'http://calypso.localhost:3000/api/oauth/token'
```

Comment:
Ideally, yes, though our current `hostname` entry can't tell if it needs the :3000 port.